// @genui/server/src/tool-executor.ts
// ============================================================
// GenUI Widget Platform - Domain Tool Executor
// Resilient API calls with Retry Loop & Mock Data Fallbacks
// ============================================================

import type { DomainToolExecutor } from './types.js';

export class OpenMeteoExecutor implements DomainToolExecutor {
  async execute(toolPath: string, params: Record<string, any>): Promise<any> {
    console.log(`Executing domain tool: ${toolPath} with params:`, params);

    try {
      switch (toolPath) {
        case '/tool/weather/current-location':
          return {
            latitude: 12.9716,
            longitude: 77.5946,
            name: 'Bangalore',
            country: 'India',
            timezone: 'Asia/Kolkata',
          };

        case '/tool/weather/geocode':
        case '/tool/location/geocode': {
          if (!params.name) {
            throw new Error('Geocoding requires a name parameter');
          }
          const count = params.count || 5;
          const language = params.language || 'en';
          const url = `https://geocoding-api.open-meteo.com/v1/search?name=${encodeURIComponent(params.name)}&count=${count}&language=${language}`;
          return await this.fetchJsonWithRetry(url);
        }

        case '/tool/weather/forecast': {
          const { latitude, longitude, ...rest } = params;
          if (latitude === undefined || longitude === undefined) {
            throw new Error('Forecast requires latitude and longitude');
          }

          const queryParams = new URLSearchParams();
          queryParams.append('latitude', latitude.toString());
          queryParams.append('longitude', longitude.toString());

          for (const [key, val] of Object.entries(rest)) {
            if (Array.isArray(val)) {
              queryParams.append(key, val.join(','));
            } else if (val !== undefined && val !== null) {
              queryParams.append(key, val.toString());
            }
          }

          if (rest.daily && !rest.timezone) {
            queryParams.append('timezone', 'auto');
          }

          const url = `https://api.open-meteo.com/v1/forecast?${queryParams.toString()}`;
          return await this.fetchJsonWithRetry(url);
        }

        case '/tool/weather/air-quality': {
          const { latitude, longitude, ...rest } = params;
          if (latitude === undefined || longitude === undefined) {
            throw new Error('Air quality requires latitude and longitude');
          }

          const queryParams = new URLSearchParams();
          queryParams.append('latitude', latitude.toString());
          queryParams.append('longitude', longitude.toString());

          for (const [key, val] of Object.entries(rest)) {
            if (Array.isArray(val)) {
              queryParams.append(key, val.join(','));
            } else if (val !== undefined && val !== null) {
              queryParams.append(key, val.toString());
            }
          }

          const url = `https://air-quality-api.open-meteo.com/v1/air-quality?${queryParams.toString()}`;
          return await this.fetchJsonWithRetry(url);
        }

        case '/tool/weather/historical': {
          const { latitude, longitude, start_date, end_date, ...rest } = params;
          if (latitude === undefined || longitude === undefined || !start_date || !end_date) {
            throw new Error('Historical weather requires latitude, longitude, start_date, and end_date');
          }

          const queryParams = new URLSearchParams();
          queryParams.append('latitude', latitude.toString());
          queryParams.append('longitude', longitude.toString());
          queryParams.append('start_date', start_date);
          queryParams.append('end_date', end_date);

          for (const [key, val] of Object.entries(rest)) {
            if (Array.isArray(val)) {
              queryParams.append(key, val.join(','));
            } else if (val !== undefined && val !== null) {
              queryParams.append(key, val.toString());
            }
          }

          if (rest.daily && !rest.timezone) {
            queryParams.append('timezone', 'auto');
          }

          const url = `https://archive-api.open-meteo.com/v1/archive?${queryParams.toString()}`;
          return await this.fetchJsonWithRetry(url);
        }

        default:
          throw new Error(`Domain tool not implemented: ${toolPath}`);
      }
    } catch (err: any) {
      console.warn(`WARNING: Resilient Tool Executor caught error for path "${toolPath}". Returning fallback mock data. Error details:`, err.message || err);
      return this.getFallbackMockData(toolPath, params);
    }
  }

  /**
   * Performs fetch request with retries and exponential backoff
   */
  private async fetchJsonWithRetry(url: string, retries = 3, delay = 500): Promise<any> {
    for (let attempt = 1; attempt <= retries; attempt++) {
      try {
        console.log(`Fetching API: ${url} (Attempt ${attempt}/${retries})`);
        const res = await fetch(url);
        
        if (res.status === 503 || res.status === 429) {
          // If service is overloaded or rate limited, wait and retry
          if (attempt === retries) {
            throw new Error(`Open-Meteo API service unavailable (status ${res.status}) after ${retries} attempts`);
          }
          const backoffDelay = delay * Math.pow(2, attempt - 1);
          console.log(`Open-Meteo responded with status ${res.status}. Backing off for ${backoffDelay}ms...`);
          await new Promise(resolve => setTimeout(resolve, backoffDelay));
          continue;
        }

        if (!res.ok) {
          const errText = await res.text().catch(() => '');
          throw new Error(`API call failed with status ${res.status}: ${errText || res.statusText}`);
        }

        return await res.json();
      } catch (err) {
        if (attempt === retries) throw err;
        const backoffDelay = delay * Math.pow(2, attempt - 1);
        await new Promise(resolve => setTimeout(resolve, backoffDelay));
      }
    }
  }

  /**
   * Returns clean, structured mock data fallbacks matching expected schemas on API failure
   */
  private getFallbackMockData(toolPath: string, params: Record<string, any>): any {
    const defaultCoords = {
      latitude: params.latitude || 12.9716,
      longitude: params.longitude || 77.5946,
      name: params.name || 'Bangalore',
      country: 'India',
      timezone: params.timezone || 'Asia/Kolkata'
    };

    switch (toolPath) {
      case '/tool/weather/geocode':
      case '/tool/location/geocode':
        return {
          results: [
            {
              id: 99999,
              name: params.name || 'Delhi',
              latitude: params.name?.toLowerCase() === 'delhi' ? 28.6519 : defaultCoords.latitude,
              longitude: params.name?.toLowerCase() === 'delhi' ? 77.2314 : defaultCoords.longitude,
              timezone: params.name?.toLowerCase() === 'delhi' ? 'Asia/Kolkata' : defaultCoords.timezone,
              country: 'Fallback Land',
              admin1: 'State'
            }
          ]
        };

      case '/tool/weather/forecast': {
        const now = new Date();
        const days = params.forecast_days || 7;
        const dailyTime: string[] = [];
        for (let i = 0; i < days; i++) {
          const d = new Date();
          d.setDate(now.getDate() + i);
          dailyTime.push(d.toISOString().split('T')[0]);
        }

        const hourlyTime: string[] = [];
        for (let i = 0; i < 24; i++) {
          const d = new Date();
          d.setHours(now.getHours() + i, 0, 0, 0);
          hourlyTime.push(d.toISOString().substring(0, 16));
        }

        const fallback: Record<string, any> = {
          latitude: defaultCoords.latitude,
          longitude: defaultCoords.longitude,
          timezone: defaultCoords.timezone,
          elevation: 100,
        };

        if (params.current) {
          fallback.current_units = {
            temperature_2m: '°C',
            weather_code: 'wmo_code',
            precipitation: 'mm'
          };
          fallback.current = {
            time: now.toISOString().substring(0, 16),
            temperature_2m: 24.5,
            weather_code: 2, // Partly cloudy
            precipitation: 0.0
          };
        }

        if (params.daily) {
          fallback.daily_units = {
            time: 'iso8601',
            temperature_2m_max: '°C',
            temperature_2m_min: '°C',
            precipitation_probability_max: '%'
          };
          fallback.daily = {
            time: dailyTime,
            weather_code: dailyTime.map((_, i) => i === 3 ? 61 : (i % 2 === 0 ? 0 : 2)), // Day 4 is rainy, others clear or cloudy
            temperature_2m_max: dailyTime.map((_, i) => 28.0 + (i % 3)),
            temperature_2m_min: dailyTime.map((_, i) => 18.0 - (i % 2)),
            precipitation_probability_max: dailyTime.map((_, i) => i === 3 ? 85 : (i % 2 === 0 ? 0 : 25))
          };
        }

        if (params.hourly) {
          fallback.hourly_units = {
            time: 'iso8601',
            temperature_2m: '°C',
            precipitation_probability: '%'
          };
          fallback.hourly = {
            time: hourlyTime,
            temperature_2m: hourlyTime.map(() => 22.0),
            weather_code: hourlyTime.map(() => 2),
            precipitation_probability: hourlyTime.map(() => 15)
          };
        }

        return fallback;
      }

      case '/tool/weather/air-quality': {
        const now = new Date();
        const hourlyTime: string[] = [];
        for (let i = 0; i < 24; i++) {
          const d = new Date();
          d.setHours(now.getHours() + i, 0, 0, 0);
          hourlyTime.push(d.toISOString().substring(0, 16));
        }

        return {
          latitude: defaultCoords.latitude,
          longitude: defaultCoords.longitude,
          current: params.current ? {
            time: now.toISOString().substring(0, 16),
            us_aqi: 45,
            pm2_5: 11.2,
            pm10: 19.5
          } : undefined,
          hourly: params.hourly ? {
            time: hourlyTime,
            us_aqi: hourlyTime.map(() => 45),
            pm2_5: hourlyTime.map(() => 11.2),
            pm10: hourlyTime.map(() => 19.5)
          } : undefined
        };
      }

      default:
        // Generic fallback object
        return {
          latitude: defaultCoords.latitude,
          longitude: defaultCoords.longitude,
          error: 'Returned mock fallback due to API failure'
        };
    }
  }
}
